cmd/nsm-init/nsm-init.go

```diff
@@ -0,0 +1,97 @@
+// Copyright (c) 2018 Cisco and/or its affiliates.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at:
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"context"
+	"flag"
+	"net"
+	"os"
+	"time"
+
+	"github.com/vishvananda/netns"
+
+	"github.com/golang/glog"
+	"github.com/ligato/networkservicemesh/pkg/nsm/apis/nsmconnect"
+	"github.com/vishvananda/netlink"
+	"google.golang.org/grpc"
+)
+
+const (
+	// clientConnectionTimeout defines time the client waits for establishing connection with the server
+	clientConnectionTimeout = time.Second * 60
+)
+
+var (
+	clientSocket = flag.String("nsm-socket", "/var/lib/networkservicemesh/nsm.ligato.io.sock", "Location of NSM process client access socket")
+)
+
+func dial(ctx context.Context, unixSocketPath string) (*grpc.ClientConn, error) {
+	c, err := grpc.DialContext(ctx, unixSocketPath, grpc.WithInsecure(), grpc.WithBlock(),
+		grpc.WithDialer(func(addr string, timeout time.Duration) (net.Conn, error) {
+			return net.DialTimeout("unix", addr, timeout)
+		}),
+	)
+
+	return c, err
+}
+
+func main() {
+	flag.Parse()
+	flag.Set("logtostderr", "true")
+
+	// Checking if NSM Client socket exists and of not crash init container
+	_, err := os.Stat(*clientSocket)
+	if err != nil {
+		glog.Errorf("NSM Client: Failure to access NSM socket at %s with error: %+v, existing...", *clientSocket, err)
+		os.Exit(1)
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), clientConnectionTimeout)
+	conn, err := dial(ctx, *clientSocket)
+	if err != nil {
+		glog.Errorf("NSM Client: Failure to communicate with the socket %s with error: %+v", *clientSocket, err)
+		os.Exit(1)
+	}
+	nsmClient := nsmconnect.NewClientConnectionClient(conn)
+	defer conn.Close()
+	defer cancel()
+	glog.Infof("NSM Client: Connection to NSM server on socket: %s succeeded.", *clientSocket)
+	glog.Infof("NSM Client: Client API %+v", nsmClient)
+	// Init related activities start here
+
+	currentNamespace, err := netns.Get()
```

Comment:
Nice find on the netns package :)  Have you tried it in a container yet?

Reply:
```
I0627 18:19:19.921610       1 nsm-init.go:71] NSM Client: Connection to NSM server on socket: /var/lib/networkservicemesh/nsm.ligato.io.sock succeeded.
I0627 18:19:19.921639       1 nsm-init.go:72] NSM Client: Client API &{cc:0xc420186000}
I0627 18:19:19.921838       1 nsm-init.go:80] NSM Client: Pod's namespace is [NS(5: 3, 4026532392)]
I0627 18:19:19.922287       1 nsm-init.go:90] NSM Client: Pod's interfaces:
I0627 18:19:19.922301       1 nsm-init.go:92]           Name: lo Type: device
I0627 18:19:19.922308       1 nsm-init.go:92]           Name: eth0 Type: veth
I0627 18:19:19.922316       1 nsm-init.go:95] NSM Client: Initialization is completed successfully, exiting...
```

Reply:
It is from the nsm init container ;)

Reply:
@sbezverk Woot!

Reply:
If there is anything else you would like me to try to do within a container, please let me know. Something that came to my mind, was creating additional namespace and for now creating new loopback there. Maybe you could suggest something more useful as I have limited imagination ;)

Reply:
@sbezverk I would not suggest creating new namespaces.

I would suggest arranging to feed it from a ConfigMap.  Users are going to want to be able to specify "I want this Network Service plugged into this interface name, etc, etc" and ConfigMaps are a good way to do that :)